internal/merkle/proof.go

```diff
@@ -0,0 +1,95 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+package merkle
+
+import (
+	"errors"
+
+	"github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// CreateProofs creates proofs for all the leaves of the binary Merkle tree
+// with the given height. It returns the root hash and the siblings matrix,
+// in bottom-up order.
+//
+// If the number of leaves exceeds the capacity for the given height,
+// an error is returned.
+func CreateProofs(leaves []model.Hash, height uint) (model.Hash, [][]model.Hash, error) {
+	pristineNode := model.Hash{}
+
+	currentLevel := leaves
+	nextLevel := make([]model.Hash, (len(leaves)+1)/2)
+	siblings := make([][]model.Hash, len(leaves))
+
+	// for each level in the tree, starting from the leaves
+	var levelIdx uint
+	for levelIdx = 0; levelIdx < height; levelIdx++ {
+		calculateSiblings(levelIdx, height, currentLevel, siblings, pristineNode)
+		calculateParents(currentLevel, nextLevel, pristineNode)
+
+		aux := currentLevel
+		currentLevel = nextLevel
+		// re-slices current level to avoid allocating multiple arrays
+		nextLevel = aux[:(len(currentLevel)+1)/2]
+		pristineNode = crypto.Keccak256Hash(pristineNode[:], pristineNode[:])
+	}
+
+	// in the end, current level is the root level
+	if len(currentLevel) > 1 {
+		return model.Hash{}, nil, errors.New("too many leaves for height")
+	}
+
+	return at(currentLevel, 0, pristineNode), siblings, nil
+}
+
+// calculateSiblings iterates over each leaf and populates the siblings matrix
+// with the appropriate sibling nodes at the current level.
+//
+// If the current level has an odd number of nodes a pristine node will be used.
+func calculateSiblings(
+	levelIdx, height uint,
+	currentLevel []model.Hash,
+	siblings [][]model.Hash,
+	pristineNode model.Hash,
+) {
+	// for each leaf
+	for idx := range siblings {
+		// creates the sibling slice if needed
+		if siblings[idx] == nil {
+			siblings[idx] = make([]model.Hash, height)
+		}
+
+		// the sibling index is to the left of the parent if leaf index is odd
+		// or to the right if leaf index is even
+		siblingIdx := (idx / 2) ^ 1
+		if levelIdx == 0 {
+			// in the leaf level, use the leaf index directly
+			siblingIdx = idx ^ 1
+		}
+
+		siblings[idx][levelIdx] = at(currentLevel, uint(siblingIdx), pristineNode)
+	}
+}
+
+// calculateParents computes the parent nodes for the next level in the Merkle
+// tree by hashing their children with the Keccak-256 hash function.
+//
+// If the current level has an odd number of nodes a pristine node will be used.
+func calculateParents(currentLevel, nextLevel []model.Hash, pristineNode model.Hash) {
+	// for each parent node
+	for idx := range nextLevel {
+		leftChild := currentLevel[2*idx]
```

Comment:
I *think* both these definitions create copies, right? Shouldn't you use `[:]` to avoid that? Or am I mistaken?

Reply:
I see this pattern elsewhere in the code. Or is it copy-on-write? Do you know?

Reply:
That's true. It seems that the `model.Hash` type is, in fact, just an array of 32 bytes. So, maybe `[:]` is necessary to avoid copying hashes around? I don't know much about Golang.

Reply:
> I _think_ both these definitions create copies, right?

Yes, they're both copies.

> Shouldn't you use `[:]` to avoid that? Or am I mistaken?

I could. But since I need to modify the `at` function to use pointers for the same reason (avoiding copies), I would rather use pointers directly.

Reply:
Don't know much about Golang, but why pointers? I realize there is no concept of a const slice (go figure...).

Reply:
> I realize there is no concept of a const slice (go figure...).
AFAIK in Golang arrays are meant to be fixed and slices are meant to be dynamic. That is just the way it is 🤷🏽‍♂️ 

> Don't know much about Golang, but why pointers?
It is the only way we can avoid copies In the current situation (a slice of byte arrays). It didn't look so bad in the end. Done at a66d705dc0b807351ce69d39378a031e3df117e3.

Reply:
When you assign a reslice of an array to a variable, you don’t create a copy, unless I’m crazy.

Reply:
My previous comment wasn't completely true. I could also re-slice it to avoid a copy as you originally proposed, but then I would end up with a `[]byte` slice, instead of a `model.Hash`. This would defeat the purpose of having a custom type to represent hashes so I went with a pointer to `model.Hash` instead.

Reply:
To be fair, that is exactly what the implementation of `crypto.Keccak256Hash`. This seems to be idiomatic, no? I am not a Go specialist.

Reply:
I agree with you. Just because I needed to use pointers for the `at` function it doesn't mean I have to use them here also when there's a more idiomatic way. I changed it at 9eb81fc6e7b4c11d24a4649bf7823edaa1c6c05a.